rln/src/protocol.rs

```diff
@@ -103,121 +171,90 @@ pub enum ProofError {
-        (
-            name.to_string(),
-            values.iter().copied().map(Into::into).collect::<Vec<_>>(),
-        )
-    });
-
+    mut builder: CircomBuilder<Bn254>,
+    proving_key: &ProvingKey<Bn254>,
+    rln_witness: RLNWitnessInput,
+) -> Result<(Proof, Vec<Fr>), ProofError> {
     let now = Instant::now();
 
-    let full_assignment = WITNESS_CALCULATOR
-        .clone()
-        .calculate_witness_element::<Bn254, _>(inputs, false)
-        .map_err(ProofError::WitnessError)?;
+    builder.push_input(
+        "identity_secret",
+        BigInt::parse_bytes(rln_witness.identity_secret.as_bytes(), 10).unwrap(),
+    );
 
-    println!("witness generation took: {:.2?}", now.elapsed());
+    for v in rln_witness.path_elements.iter() {
+        builder.push_input(
+            "path_elements",
+            BigInt::parse_bytes(v.as_bytes(), 10).unwrap(),
+        );
+    }
 
-    let mut rng = thread_rng();
-    let rng = &mut rng;
+    for v in rln_witness.identity_path_index.iter() {
+        builder.push_input("identity_path_index", BigInt::from(*v));
+    }
+
+    builder.push_input(
+        "x",
+        BigInt::parse_bytes(rln_witness.x.as_bytes(), 10).unwrap(),
+    );
+
+    builder.push_input(
+        "epoch",
+        BigInt::parse_bytes(rln_witness.epoch.strip_prefix("0x").unwrap().as_bytes(), 16).unwrap(),
+    );
+
+    builder.push_input(
+        "rln_identifier",
+        BigInt::parse_bytes(rln_witness.rln_identifier.as_bytes(), 10).unwrap(),
+    );
 
-    let r = ark_bn254::Fr::rand(rng);
-    let s = ark_bn254::Fr::rand(rng);
+    let circom = builder.build().unwrap();
+
+    // Get the populated instance of the circuit with the witness
+    let inputs = circom.get_public_inputs().unwrap();
+
+    println!("witness generation took: {:.2?}", now.elapsed());
 
     let now = Instant::now();
 
-    let ark_proof = create_proof_with_reduction_and_matrices::<_, CircomReduction>(
-        &ZKEY.0,
-        r,
-        s,
-        &ZKEY.1,
-        ZKEY.1.num_instance_variables,
-        ZKEY.1.num_constraints,
-        full_assignment.as_slice(),
-    )?;
+    // Generate a random proof
+    let mut rng = thread_rng();
+
+    let ark_proof = create_random_proof_with_reduction::<_, _, _, CircomReduction>(
+        circom,
+        proving_key,
+        &mut rng,
+    )
+    .unwrap();
+
     let proof = ark_proof.into();
+
     println!("proof generation took: {:.2?}", now.elapsed());
 
-    Ok(proof)
+    Ok((proof, inputs))
 }
 
-// TODO Update API here
-
-/// Verifies a given semaphore proof
+/// Verifies a given RLN proof
 ///
 /// # Errors
 ///
 /// Returns a [`ProofError`] if verifying fails. Verification failure does not
 /// necessarily mean the proof is incorrect.
 pub fn verify_proof(
-    root: Field,
-    nullifier_hash: Field,
-    signal_hash: Field,
-    external_nullifier_hash: Field,
-    proof: &Proof,
+    verifying_key: &VerifyingKey<Bn254>,
+    proof: Proof,
+    inputs: Vec<Fr>,
```

Comment:
Why is this <Fr> instead of Field?

I don't quite see the purpose of removing the named parameters here as that removes context, ark_verify_proof is already general, the purpose of this function is to be specific to RLN AFAICT

Reply:
You're right. I created a new structure which is RLNProofValues and I now pass this to verify_proof instead of inputs: Vec<Fr>. This goes into the direction to have generate_proof and verify_proof have a general API, where to the first one you pass a module-specific witness (RLNWitness) while to the second proof-values to be verified (RLNProofValues).